lib/OpenQA/Task/Git/Clone.pm

```diff
@@ -0,0 +1,138 @@
+# Copyright SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+package OpenQA::Task::Git::Clone;
+use Mojo::Base 'Mojolicious::Plugin', -signatures;
+use Mojo::Util 'trim';
+
+use OpenQA::Utils qw(run_cmd_with_log_return_error);
+use Cwd 'abs_path';
+use Time::Seconds 'ONE_HOUR';
+
+sub register ($self, $app, @) {
+    $app->minion->add_task(git_clone => \&_git_clone);
+}
+
+
+# $clones is a hashref with paths as keys and urls to git repos as values.
+# The urls may also refer to a branch via the url fragment.
+# If no branch is set, the default branch of the remote (if target path doesn't exist yet)
+# or local checkout is used.
+# If the target path already exists, an update of the specified branch is performed,
+# if the url matches the origin remote.
+
+sub _git_clone ($job, $clones) {
+    my $app = $job->app;
+    my $job_id = $job->id;
+
+    # Prevent multiple git clone tasks for the same path to run in parallel
+    my @guards;
+    for my $path (sort keys %$clones) {
+        while (-l $path) {
+            # guard symlink target
+            my $newpath = abs_path($path);
+            last if ($newpath eq $path);
+            $path = $newpath;
+        }
+        my $guard = $app->minion->guard("git_clone_${path}_task", 2 * ONE_HOUR);
+        return $job->retry({delay => 30 + int(rand(10))}) unless $guard;
+        push(@guards, $guard);
+    }
+
+    my $log = $app->log;
+    my $ctx = $log->context("[#$job_id]");
+
+    # iterate clones sorted by path length to ensure that a needle dir is always cloned after the corresponding casedir
+    for my $path (sort { length($a) > length($b) } keys %$clones) {
+        my $url = $clones->{$path};
+        die("Don't even think about putting '..' into '$path'.") if $path =~ /\.\./;
```

Comment:
```suggestion
        die "Don't even think about putting '..' into '$path'." if $path =~ /\.\./;
```

Reply:
@asdil12 This is unresolved. It is only visible in the **Files changed** tab